### ISSUE TITLE
Command alias rework

### ISSUE THREAD

**Issue by SOF3** [pmmp/PocketMine-MP#3371]
# Motivation
The current command system is largely derived from bukkit, suffering from the following problems:
- to customize command names, plugin has to make their own config entry and register the command dynamically
- although alias can be set in pocketmine.yml, it is often unknown to most users, and even if it is, managing a patch layer on top of names and aliases registered by plugins is a pain since it is unclear what gets used and what doesn't
- users cannot unlink an alias from a command (for whatever reason)
- plugins trying to replace a command from the core or another plugin might suffer from undefined behavior due to racing overwrites
- it is not trivial for plugins to remap command names dynamically without implementing their own CommandMap
- plugins with the same command name race to use an alias space with undefined behavior

This proposal aims to solve these problems by redefining the concepts of command name and aliases and creating an alternative alias configuration. (Sounds orthogonal, but they are closely related)

## Proposal
The command name defined by plugins is regarded as a semi-internal identifier for a command (at a slightly higher level than function names). The command name is not expected to be used at user interface, but can still be referred to with `pluginname:commandname` (this is also the full name to identify the command externally).

A command alias is a way for a user to access a command through chat window. The command name is an implicit alias for the command, along with extra aliases as specified by the command.

When a plugin registers a command, PocketMine tracks if this command (as identified by its full name) was previously registered in previous runs. If it was never registered, for each alias, PocketMine overwrites the alias' entry in a file `$DATA_PATH/commands.yml` to this command full name. PocketMine also maintains a map [full name => Command object].

The AliasMap interface is introduced, with the single method `getCommand(string $name): ?Command`.

The only builtin implementation of AliasMap is FileAliasMap, which retrieves $name from the parameterized file (specified as commands.yml in server initialization code), then resolving the full name into the Command object.

The Player class (see Unresolved section) exposes getset for an aliasMap field, which is used to resolve the command when a player executed from the chat window. It is by default initialized with FileAliasMap("$DATA_PARH/commands.tml"). This can be set by olugins to alternative implementations, such as ones that delegate to player-specific files or world-specific files. This field should be settable under all reasonable states.

# Unresolved
- How should other CommandSender implementations behave? Should they also use an AliasMap, or is it just specific to Player?
- Should PocketMine warn/error/ignore/continue if a plugin is registering an alias that overwrites an existing one?

**Comment by dktapps** [comment]
tl;dr: you want a config file which allows command aliases to be edited, remapped and removed, and not just added.

**Comment by SOF3** [comment]
Another main point is that I don't want the command name to have any special status over ordinary aliases, other than usage in the `pluginname:commandname` qualifier (but not alone).

**Comment by SOF3** [comment]
TL;DR: The flow is like this

- plugin "foo" enables
- plugin "foo" tells pocketmine that we supply a command called "bar"
- plugin "foo" tells pocketmine that we suggest the user to alias "bar" with "qux" and "corge" as well
- pocketmine detects that this is the first time "foo" supplies "bar"
- pocketmine adds the entries `bar: foo:bar`, `qux: foo:bar`, `corge: foo:bar` into a commands.yml
- user can edit commands.yml to delete some lines. plugins can't add the same command to commands.yml again after the user deleted it manually.

**Comment by dktapps** [comment]
After further investigation, I realized that the `aliases` config of `pocketmine.yml` already permits adding, remapping and removing aliases.

For example, if two plugins have the alias `addgroup`, the user can do the following:

```yml
    addgroup: [pureperms:addgroup]
```
or
```yml
    addgroup: [otherperms:addgroup]
```
to remap the alias, or
```yml
    addgroup: []
```
to remove it altogether.

**Comment by SOF3** [comment]
Users can still access the command using `pureperms:addgroup`. (But why does that even exist anyway?)

Also it is unclear what aliases are provided by each plugin, since pocketmine.yml is not dynamically updated.

**Comment by dktapps** [comment]
I've been mulling this over in the last few days trying to picture how it would look when implemented.

One unresolved question in the RFC is what to do when a plugin providing a certain command is removed. e.g. `foo:bar` is no longer present, so what should its aliases do?
a) become free for reuse by another plugin?
b) stay reserved and blackhole the command?

This question is of particular importance because it influences how the system would need to be designed for plugin loading. For example:
- plugin A registers `a:foo` and plugin B registers `b:foo`, in that order
- commands.yml remembers that `foo` = `a:foo`
- on the next run, plugin B loads first, but `foo` is already reserved for `a:foo`

In this scenario, if we go with option a), the resolution of plugin aliases would need to be deferred until all plugins are loaded and enabled, to ensure the command the alias is pointing at actually exists. b) would be simpler (just never register if the alias is already taken, whether or not the target exists), but I'm not sure if that's a good idea for UX.

**Comment by dktapps** [comment]
For a config file like this to work well, we would also need to mandate that command namespaces comply with some central standard (e.g. plugin name), otherwise users would get confused. I'm not really sure how to do this without binding the `CommandMap` API to `plugin` package code.

**Comment by dktapps** [comment]
#6685 implements the internal changes needed for this. Storage is left up to plugins for now.
